rllib/models/experimental/torch/encoder.py

```diff
@@ -142,26 +138,130 @@ def forward(self, inputs: TensorDict, **kwargs) -> ForwardOutputType:
+    @override(Model)
+    def _forward(self, inputs: NestedDict, **kwargs) -> NestedDict:
+        return {ENCODER_OUT: inputs[SampleBatch.OBS], STATE_OUT: inputs[STATE_IN]}
+
+
+class TorchActorCriticEncoder(TorchModel, Encoder):
+    """An encoder that potentially holds two encoders.
+
+    This is a special case of encoder that potentially holds two encoders:
+    One for the actor and one for the critic. If not, it will use the same encoder
+    for both. The two encoders are of the same type and we can therefore make the
+    assumption that they have the same input and output specs.
+    """
+
+    def __init__(self, config: ModelConfig) -> None:
+        TorchModel.__init__(self, config)
+        Encoder.__init__(self, config)
+
+        if self.config.shared:
+            self.encoder = self.config.base_encoder_config.build(framework="torch")
+        else:
+            self.actor_encoder = self.config.base_encoder_config.build(
+                framework="torch"
+            )
+            self.critic_encoder = self.config.base_encoder_config.build(
+                framework="torch"
+            )
+
+        if self.config.shared:
+            self.input_spec = self.encoder.input_spec
+        else:
+            actor_input_spec = self.actor_encoder.input_spec
+            critic_input_spec = self.critic_encoder.input_spec
+            assert actor_input_spec == critic_input_spec
+            # We only make assumptions about OBS, STATE_IN and SEQ_LENS here.
+            # By extending this class, one could feed inputs that hold more keys.
+
+            self.input_spec = SpecDict(
+                {
+                    SampleBatch.OBS: actor_input_spec[SampleBatch.OBS],
+                    STATE_IN: {
+                        ACTOR: actor_input_spec[STATE_IN],
+                        CRITIC: critic_input_spec[STATE_IN],
+                    },
+                    SampleBatch.SEQ_LENS: actor_input_spec[SampleBatch.SEQ_LENS],
+                }
+            )
+
+        if self.config.shared:
+            state_out_spec = self.encoder.output_spec[STATE_OUT]
+            actor_out_spec = self.encoder.output_spec[ENCODER_OUT]
+            critic_out_spec = self.encoder.output_spec[ENCODER_OUT]
+        else:
+            state_out_spec = {
+                ACTOR: self.actor_encoder.output_spec[STATE_OUT],
+                CRITIC: self.critic_encoder.output_spec[STATE_OUT],
+            }
+            actor_out_spec = self.actor_encoder.output_spec[ENCODER_OUT]
+            critic_out_spec = self.critic_encoder.output_spec[ENCODER_OUT]
+            assert actor_out_spec == critic_out_spec
+
+        self.output_spec = {
+            ENCODER_OUT: {ACTOR: actor_out_spec, CRITIC: critic_out_spec},
+            STATE_OUT: state_out_spec,
+        }
+
+    @override(Model)
+    def get_initial_state(self):
+        if self.config.shared:
+            return self.encoder.get_initial_state()
+        else:
+            return {
```

Comment:
I don't think we need to support this in the basic implementation. When you use RNNs in an actor crtitic algorithm I think it's expected that the latent temporal model is shared between the two. So we are creating too many features here. We should simply say hey you cannot use non shared encoders if lstm = True

Reply:
Got ya! Agreed :)

Reply:
It's kind of wonky to assert this inside this encoder though because the beauty of this one is, I think, that it does not care for the underlying model's details. At this point we already support this in the basic implementation so unless it leads to issues I believe we should just leave it there for now.

Reply:
Having a few RNN states running is a bit wonky imo.